sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py

```diff
@@ -390,3 +395,193 @@ def compute(
             "Only 64bit float datasets are supported at this time, "
             f"got: X.dtype={X.dtype} and Y.dtype={Y.dtype}."
         )
+
+
+class PairwiseDistancesArgKminLabels(PairwiseDistancesReduction):
+    """Compute the argkmin of row vectors of X on the ones of Y with labels.
+
+    For each row vector of X, computes the indices of k first the rows
+    vectors of Y with the smallest distances. Computes weighted mode of labels.
+
+    PairwiseDistancesArgKminLabels is typically used to perform
+    bruteforce k-nearest neighbors queries when the weighted mode of the labels
+    for the k-nearest neighbors are required, such as in `predict` methods.
+
+    This class is not meant to be instanciated, one should only use
+    its :meth:`compute` classmethod which handles allocation and
+    deallocation consistently.
+    """
```

Comment:
> As a first prototype, such a back-end (say `KNeighborsClassifierProbaPrediction`) could solely return a (say) `proba` array of shape `(n_samples, n_classes)` (or another axis ordering) which could be then reduced using `np.argmax` on the last axe in Python.

 This sounds good -- originally I didn't know how I wanted to handle it either (e.g. how much overlap could we establish between classification/regression) but starting with something tractable is best. I'll go with your suggestion and defer the `argmax` to the Python layer. Similarly, we don't need to do anything special for multi-output since it is exactly equivalent to a series of single-output calls. Since we're already saturating parallelism at the lower levels, doing this in sequence should be fine.

Reply:
> This sounds good -- originally I didn't know how I wanted to handle it either (e.g. how much overlap could we establish between classification/regression) but starting with something tractable is best. I'll go with your suggestion and defer the argmax to the Python layer.

:+1: 

> Similarly, we don't need to do anything special for multi-output since it is exactly equivalent to a series of single-output calls. Since we're already saturating parallelism at the lower levels, doing this in sequence should be fine.

Doing them in sequence is the way to go, put in the core of the parallel computations (see my previous comment).